src/Platforms/Keil/UtestPlatform.cpp

```diff
@@ -0,0 +1,218 @@
+            function(data);
+            jmp_buf_index--;
+            return 1;
+        }
+        return 0;
+    }
+
+    static void PlatformSpecificLongJmpImplementation()
+    {
+        jmp_buf_index--;
+        longjmp(test_exit_jmp_buf[jmp_buf_index], 1);
+    }
+
+    static void PlatformSpecificRestoreJumpBufferImplementation()
+    {
+        jmp_buf_index--;
+    }
+
+    void (*PlatformSpecificLongJmp)() = PlatformSpecificLongJmpImplementation;
+    int (*PlatformSpecificSetJmp)(void (*function)(void*), void*) = PlatformSpecificSetJmpImplementation;
+    void (*PlatformSpecificRestoreJumpBuffer)() = PlatformSpecificRestoreJumpBufferImplementation;
+
+
+
+    ///////////// Time in millis
+    /*
+    *  In Keil MDK-ARM, clock() default implementation used semihosting.
+    *  Resolutions is user adjustable (1 ms for now)
+    */
+    static long TimeInMillisImplementation()
+    {
+        clock_t t = clock();
+        
+       t = t * 10;
+       
+        return t;
+    }
+
+    long (*GetPlatformSpecificTimeInMillis)() = TimeInMillisImplementation;
+
+    static const char* TimeStringImplementation()
+    {
+        time_t tm = 0;//time(NULL); // todo
+        return ctime(&tm);
+    }
+
+    const char* (*GetPlatformSpecificTimeString)() = TimeStringImplementation;
+
+    int PlatformSpecificAtoI(const char* str)
+    {
+        return atoi(str);
+    }
+
+    /* The ARMCC compiler will compile this function with C++ linkage, unless
+     * we specifically tell it to use C linkage again, in the function definiton.
+     */
+    extern int (*PlatformSpecificVSNprintf)(char *str, size_t size, const char* format, va_list args) = vsnprintf;
+
+    static PlatformSpecificFile PlatformSpecificFOpenImplementation(const char* filename, const char* flag)
+    {
+       return 0;
+    }
+
+    static void PlatformSpecificFPutsImplementation(const char* str, PlatformSpecificFile file)
+    {
+        printf("%s", str);
+    }
+
+    static void PlatformSpecificFCloseImplementation(PlatformSpecificFile file)
+    {
+    }
+
+    static void PlatformSpecificFlushImplementation()
+    {
+    }
+
+    PlatformSpecificFile (*PlatformSpecificFOpen)(const char*, const char*) = PlatformSpecificFOpenImplementation;
+    void (*PlatformSpecificFPuts)(const char*, PlatformSpecificFile) = PlatformSpecificFPutsImplementation;
+    void (*PlatformSpecificFClose)(PlatformSpecificFile) = PlatformSpecificFCloseImplementation;
+
+    int (*PlatformSpecificPutchar)(int) = putchar;
+    void (*PlatformSpecificFlush)() = PlatformSpecificFlushImplementation;
+    void* (*PlatformSpecificMalloc)(size_t) = malloc;
+    void* (*PlatformSpecificRealloc) (void*, size_t) = realloc;
+    void (*PlatformSpecificFree)(void*) = free;
+    void* (*PlatformSpecificMemCpy)(void* s1, const void* s2, size_t size) = memcpy;
+    void* (*PlatformSpecificMemset)(void*, int, size_t) = memset;
+
+    static int IsNanImplementation(double d)
+    {
+#       ifdef __MICROLIB
```

Comment:
Yes __MICROLIB doesn't handle NAN and INF.

__MICROLIB is a compiler option for smaller and faster code. For my production code I will use MicroLib option. To support both variants I added this precompiler switch.

Because the compiler returns for infinity and not_a_number both 0. The StringFrom function returns in all cases "Nan - Not a number".

´´´
// SimpleString.cpp line 492
SimpleString StringFrom(double value, int precision)
{
    if (PlatformSpecificIsNan(value))
        return "Nan - Not a number";
    else if (PlatformSpecificIsInf(value))
        return "Inf - Infinity";
    else
        return StringFromFormat("%.*g", precision, value);
}
´´´